app/Platform/Actions/UpdateGameAchievementsMetricsAction.php

```diff
@@ -18,6 +19,20 @@ public function execute(Game $game): void
         $playersTotal = $game->players_total;
         $playersHardcore = $game->players_hardcore;
 
+        // If players_total is 0, calculate from actual unlocks
+        // This handles cases where metrics are updated before game player counts (ie: tests)
+        if ($playersTotal === 0) {
+            $playerCounts = PlayerAchievement::query()
+                ->whereIn('achievement_id', $game->achievements()->published()->pluck('ID'))
+                ->whereHas('user', function ($query) { $query->tracked(); })
+                ->selectRaw('COUNT(DISTINCT user_id) as total_players')
+                ->selectRaw('COUNT(DISTINCT CASE WHEN unlocked_hardcore_at IS NOT NULL THEN user_id END) as hardcore_players')
+                ->first();
+
+            $playersTotal = $playerCounts->total_players ?? 0;
+            $playersHardcore = $playerCounts->hardcore_players ?? 0;
+        }
```

Comment:
This doesn't handle the case where the player counts should be inherited from a parent set. The comment implies this is only a problem in tests, so it might not warrant the extra logic.

If I proceed to break out the player count updating logic into a separate job as we discussed earlier, this should just call into that action.